server/service/src/barcode.rs

```diff
@@ -1,25 +1,149 @@
 };
+use util::uuid::uuid;
+
+use crate::service_provider::ServiceContext;
 
 use super::{get_default_pagination, i64_to_u32, ListError, ListResult};
 
 pub const MAX_LIMIT: u32 = 5000;
 pub const MIN_LIMIT: u32 = 1;
 
-pub fn get_barcodes(
-    connection_manager: &StorageConnectionManager,
-    pagination: Option<PaginationOption>,
-    filter: Option<BarcodeFilter>,
-    sort: Option<BarcodeSort>,
-) -> Result<ListResult<Barcode>, ListError> {
-    let pagination = get_default_pagination(pagination, MAX_LIMIT, MIN_LIMIT)?;
-    let connection = connection_manager.connection()?;
-    let repository = BarcodeRepository::new(&connection);
-
-    Ok(ListResult {
-        rows: repository.query(pagination, filter.clone(), sort)?,
-        count: i64_to_u32(repository.count(filter)?),
-    })
+pub struct InsertResult {
+    pub id: String,
+    pub value: String,
+    pub item_id: String,
+    pub pack_size: Option<i32>,
+}
+
+pub struct BarcodeInput {
+    pub value: String,
+    pub item_id: String,
+    pub pack_size: Option<i32>,
+}
+
+#[derive(Debug, PartialEq)]
+pub enum InsertBarcodeError {
+    DatabaseError(RepositoryError),
+    InternalError(String),
+    BarcodeAlreadyExists,
+    InvalidItem,
+}
+impl From<RepositoryError> for InsertBarcodeError {
+    fn from(error: RepositoryError) -> Self {
+        InsertBarcodeError::DatabaseError(error)
+    }
+}
+
+pub trait BarcodeServiceTrait: Sync + Send {
+    fn get_barcode(
+        &self,
+        ctx: &ServiceContext,
+        id: String,
+    ) -> Result<Option<Barcode>, RepositoryError> {
+        let repository = BarcodeRepository::new(&ctx.connection);
+
+        Ok(repository
+            .query_by_filter(BarcodeFilter::new().id(EqualFilter::equal_to(&id)))?
+            .pop())
+    }
+
+    fn get_barcodes(
+        &self,
+        connection_manager: &StorageConnectionManager,
+        pagination: Option<PaginationOption>,
+        filter: Option<BarcodeFilter>,
+        sort: Option<BarcodeSort>,
+    ) -> Result<ListResult<Barcode>, ListError> {
+        let pagination = get_default_pagination(pagination, MAX_LIMIT, MIN_LIMIT)?;
+        let connection = connection_manager.connection()?;
+        let repository = BarcodeRepository::new(&connection);
+
+        Ok(ListResult {
+            rows: repository.query(pagination, filter.clone(), sort)?,
+            count: i64_to_u32(repository.count(filter)?),
+        })
+    }
+
+    fn insert_barcode(
+        &self,
+        ctx: &ServiceContext,
+        barcode: &BarcodeInput,
+    ) -> Result<Barcode, InsertBarcodeError> {
+        let result = ctx
+            .connection
+            .transaction_sync(|con| {
+                validate(con, &ctx.store_id, &barcode)?;
+                let barcode_repository = BarcodeRepository::new(con);
+                let new_barcode = BarcodeRow {
+                    id: uuid(),
+                    value: barcode.value.clone(),
+                    item_id: Some(barcode.item_id.clone()),
```

Comment:
can barcode exist without item ?

Reply:
yeah, in mSupply it can. so I've done the same

Reply:
@GaryWilletts do you know why this would be ?

Reply:
@mark-prins I don't think a barcode can exist independent of an item in mSupply, I can't think of anywhere in the system you could do that. It wouldn't make sense because it always has to refer to an item.

@andreievg only because Mark's been drinking wonky juice 😆

Reply:
that was based on the database schema rather than functionality. I'll change the oms schema to NOT NULL in addition to the FK. it's clearer that way.